gstreamer/src/query.rs

```diff
@@ -15,114 +15,185 @@ use std::ptr;
+where
+    Self: Sized,
+{
+    type GstType;
+
+    fn is_writable(&self) -> bool {
+        unsafe {
+            from_glib(ffi::gst_mini_object_is_writable(self.as_ptr()
+                as *const ffi::GstMiniObject))
+        }
     }
 
-    pub fn new_duration(fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_duration(fmt.to_glib())) }
+    unsafe fn as_ptr(&self) -> *const Self::GstType;
+
+    unsafe fn as_mut_ptr(&self) -> *mut Self::GstType;
+
+    unsafe fn from_glib_none(ptr: *mut Self::GstType) -> Self;
+    unsafe fn from_glib_borrow(ptr: *mut Self::GstType) -> Self;
+    unsafe fn from_glib_full(ptr: *mut Self::GstType) -> Self;
+}
+
+impl<T: MiniObject2> Deref for GstRc2<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        println!("is_writable");
+        self.0.as_ref().unwrap()
     }
+}
 
-    pub fn new_latency() -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_latency()) }
+impl<T: glib::StaticType> glib::StaticType for GstRc2<T> {
+    fn static_type() -> glib::types::Type {
+        T::static_type()
     }
+}
 
-    pub fn new_seeking(fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_seeking(fmt.to_glib())) }
+impl<T: MiniObject2 + 'static> FromGlibPtrNone<*mut T::GstType> for GstRc2<T> {
+    unsafe fn from_glib_none(ptr: *mut T::GstType) -> Self {
+        GstRc2(Some(T::from_glib_none(ptr)))
     }
+}
 
-    pub fn new_segment(fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_segment(fmt.to_glib())) }
+impl<T: MiniObject2 + 'static> FromGlibPtrBorrow<*mut T::GstType> for GstRc2<T> {
+    unsafe fn from_glib_borrow(ptr: *mut T::GstType) -> Self {
+        GstRc2(Some(T::from_glib_borrow(ptr)))
     }
+}
 
-    pub fn new_convert<V: Into<GenericFormattedValue>>(value: V, dest_fmt: ::Format) -> Self {
-        assert_initialized_main_thread!();
-        let value = value.into();
-        unsafe {
-            from_glib_full(ffi::gst_query_new_convert(
-                value.get_format().to_glib(),
-                value.get_value(),
-                dest_fmt.to_glib(),
-            ))
-        }
+impl<T: MiniObject2 + 'static> FromGlibPtrFull<*mut T::GstType> for GstRc2<T> {
+    unsafe fn from_glib_full(ptr: *mut T::GstType) -> Self {
+        GstRc2(Some(T::from_glib_full(ptr)))
     }
+}
 
-    pub fn new_formats() -> Self {
-        assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gst_query_new_formats()) }
+// *** Draft Rework ***
+// Query specific definitions start here
+glib_wrapper! {
+    pub struct QueryRef(Shared<ffi::GstQuery>);
+
+    match fn {
+        ref => |ptr| ffi::gst_mini_object_ref(ptr as *mut ffi::GstMiniObject),
+        unref => |ptr| ffi::gst_mini_object_unref(ptr as *mut ffi::GstMiniObject),
```

Comment:
Due to this, you can take a `&mut QueryRef` (as from `get_mut` e.g.) and use its `Clone` impl to get a `QueryRef` that uses the very same pointer but it has a refcount > 1. You then have a `&mut QueryRef` that is not writable anymore.